### DISCUSSION TITLE
No discharging after NodeRed Update

### DISCUSSION THREAD

**Discussion by doncuco** [sonnen-tools/node-red#12] [Ideas]
There was a NodeRed update yesterday for my home assistant. Or something else is wrong.
But right now prices are at 36cent and my battery is keeping its SOC at 34%.

Looks like if the Plugin gets restarted it looses the "calculateddischargecosts".

Could you have a look at that?

Perhaps you could optimize that so that this value gets stored somewhere an can be recovered in this case.
Or if the new calculated value for max charging price is below the actual Tibber price, then use that, so hat the battery can go to self-consumption.

Right now I set the battery manually to self-consumption mode.
But the Flow keeps setting it to manually. Could I manually overwright the Not a Number value of clculateddischargecost?

And here comes the next idea. Perhaps you could integrate a button in the dashboard to do this.

<img width="1053" alt="image" src="https://github.com/sonnen-tools/node-red/assets/122402791/68b5f148-d50b-4741-b3f6-e57b124c012c">

**Comment by markusjnagel** [comment]
Hey :wave: - yeah, during the restart you lost all your context (that stores your chargelog). Will enhance that to persist the context, so it survives a restart.

However, what shouldn't happen is that you're not seeing any Tibber prices. That is a :bug: - I can see that you have them in the system (no problem with the Tibber query) because it shows the current Tibber price.

What you can do (as a workaround) is to disable the `sonnen_discharge_control` in the Sonnenbatterie Config Node (and then hit "deploy" again)
![image](https://github.com/sonnen-tools/node-red/assets/33287465/68d94aeb-c67a-442f-a58d-8f5824d33aa5)

Depending on what state it is in when you do that, you might have to switch it to "self-consumption" manually one last time - but when that setting is off, it won't interfere anymore.

Since your chargelog is empty, it interprets that as "0" - which is cheaper than Tibber :wink: 
That is usually not a problem, because normally when you don't have a chargelog, it means that you have hit 0% - the chargelog will be cleared in that case. And since the battery is empty anyway, it doesn't matter if it is blocked for discharging.....

But, yeah, I see your problem.

**Comment by markusjnagel** [comment]
hey @doncuco - please see the [releasenotes](https://github.com/sonnen-tools/node-red/blob/main/releasenotes.md) and https://github.com/sonnen-tools/node-red/blob/main/docs/setup/nodered/context-persistence.md

If you need that (seems your home assistant is updating without asking you...), this is how you configure it. There is no way to influence that from a flow, so you have to configure it once on your Node-Red instance. 
Hope that helps :vulcan_salute: 

oh, and the other issue you had is addressed here #14 - I have already fixed that but my current development version has a few other loose ends that I need to tie together before I release 0.0.4 (WIP and "final").

**Reply by doncuco** [reply]
I have to check what exactly happened. But this solution looks good for me. Especially because my system runs in a hyper-v on an enterprise SSD with lots of TBW. So storing information is not a problem.

So then I have to figure out how to access the settings.js in node red home assistant plugin.

Very appreciate your help!

**Reply by markusjnagel** [reply]
ok, in your case you can also leave it at the default (take the example from the `settings.js`), because you don't have to worry about read/write cycles on an SDcard (no need to extend the `flushInterval`)
```
    contextStorage: {
       default: {
           module:"localfilesystem"
           },
    },

```

**Reply by TobiasW** [reply]
Hello Markus,

first of all, thanks for your great piece of software. For my part, I have solved it this way, because I use Node-Red for many things:
```
contextStorage: {
   default: "memoryOnly",
   memoryOnly: { module: 'memory' },
   file: { module: 'localfilesystem' }
}
```

and have changed all accesses to the ChargeLog as follows

```
global.set("chargeLog", ChargeLog, "file");
global.get("chargeLog", "file");
```

**Reply by doncuco** [reply]
I am running NodeRed as Home Assistant PlugIn and haven't figured out how to set this values.
An other easy way would be to simply set the battery to Auto mode if this value is unavailable.
